docs/architecture/adr-050-sign-mode-textual-annex1.md

```diff
@@ -0,0 +1,284 @@
+```
+
+we have the following value-rendered encoding:
+
+```
+Allowed messages: 2 strings
+Allowed messages (1/2): cosmos.bank.v1beta1.MsgSend
+Allowed messages (2/2): cosmos.gov.v1.MsgVote
+End of Allowed messages
+```
+
+### `message`
+
+- Applies to Protobuf messages whose name does not start with `Msg`
+  - For `sdk.Msg`s, please see [ADR-050](./adr-050-sign-mode-textual.md)
+  - alternatively, we can decide to add a protobuf option to denote messages that are `sdk.Msg`s.
+- Field names follow [sentence case](https://en.wiktionary.org/wiki/sentence_case)
+  - replace `_` with a spaces
+  - capitalize first letter of the setence
+- Field names are ordered by their Protobuf field number
+- Nesting:
+  - if a field contains a nested message, we value-render the underlying message using the template:
+  ```
+  <field_name>: <1st line of value-rendered message>
+  > <lines 2-n of value-rendered message>             // Notice the `>` prefix.
+  ```
+  - `>` character is used to denote nesting. For each additional level of nesting, add `>`.
+
+#### Examples
+
+Given the following Protobuf messages:
+
+```proto
+enum VoteOption {
+  VOTE_OPTION_UNSPECIFIED = 0;
+  VOTE_OPTION_YES = 1;
+  VOTE_OPTION_ABSTAIN = 2;
+  VOTE_OPTION_NO = 3;
+  VOTE_OPTION_NO_WITH_VETO = 4;
+}
+
+message WeightedVoteOption {
+  VoteOption option = 1;
+  string     weight = 2 [(cosmos_proto.scalar) = "cosmos.Dec"];
+}
+
+message Vote {
+  uint64 proposal_id = 1;
+  string voter       = 2 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  reserved 3;
+  repeated WeightedVoteOption options = 4;
+}
+```
+
+we get the following encoding for the `Vote` message:
+
+```
+Vote object
+> Proposal id: 4
+> Vote: cosmos1abc...def
+> Options: 2 WeightedVoteOptions
+> Options (1/2): WeightedVoteOption object
+>> Option: Yes
+>> Weight: 0.7
+> Options (2/2): WeightedVoteOption object
+>> Option: No
+>> Weight: 0.3
+> End of Options
+```
+
+### Enums
+
+- String case convention: snake case to sentence case
+- Allow optional annotation for textual name (TBD)
+- Algorithm:
+  - convert enum name (`VoteOption`) to snake_case (`VOTE_OPTION`)
+  - truncate that prefix + `_` from the enum name if it exists (`VOTE_OPTION_` gets stripped from `VOTE_OPTION_YES` -> `YES`)
+  - convert rest to sentence case: `YES` -> `Yes`
+  - in summary: `VOTE_OPTION_YES` -> `Yes`
+
+#### Examples
+
+See example above with `message Vote{}`.
+
+### `google.protobuf.Any`
+
+- Applies to `google.protobuf.Any`
+- Rendered as:
+
+```
+Object: <type_url>
+> <value rendered underlying message>
+```
+
+#### Examples
+
+```
+Object: /cosmos.gov.v1.Vote
+> Proposal id: 4
+> Vote: cosmos1abc...def
+> Options: 2 WeightedVoteOptions
+> Options (1/2): WeightedVoteOption object
+>> Option: Yes
+>> Weight: 0.7
+> Options (2/2): WeightedVoteOption object
+>> Option: No
+>> Weight: 0.3
+> End of Options
+```
+
+### `google.protobuf.Timestamp`
+
+Rendered as either ISO8601 (`2021-01-01T12:00:00Z`).
```

Comment:
I'm working on the implementation of Timestamp now.

I think what we want is [RFC 3339](https://tools.ietf.org/html/rfc3339). ISO8601 is a superset of that which does way too much for our needs. It does fun stuff like PT24H which is now until 24 hours later.

Also it is important to perseve nanosecond precision. Otherwise we loose information. Some more detail on how fractional seconds are displayed would be useful.

Reply:
Ah I see! OK that's nice, I'll update the ADR and add an example with fractional seconds.